dll/win32/shell32/wine/shellpath.c

```diff
@@ -208,29 +217,61 @@ PathQualifyExW(_Inout_ LPWSTR pszPath, _Inout_opt_ LPCWSTR pszDir, _In_ DWORD dw
                 ++pchTemp;
         }
 
-#if 0 // FIXME: IsLFNDriveW has a bug
         bLFN = IsLFNDriveW(szRoot);
-#endif
         if (!bLFN)
         {
-            /* FIXME: Support non-LFN */
+            PWCHAR pch;
+            for (pch = pchTemp; *pch != UNICODE_NULL; ++pch)
+            {
+#define PATH_CHAR_CLASS_DOT         0x00000004
+#define PATH_CHAR_CLASS_BACKSLASH   0x00000008
+#define PATH_CHAR_CLASS_COLON       0x00000010
+#define PATH_CHAR_CLASS_OTHER_VALID 0x00000100
+#define VALID_SHORT_PATH_CHAR_CLASSES ( \
+    PATH_CHAR_CLASS_DOT | \
+    PATH_CHAR_CLASS_BACKSLASH | \
+    PATH_CHAR_CLASS_COLON | \
+    PATH_CHAR_CLASS_OTHER_VALID \
+)
```

Comment:
These defines would be better in the same header as where PathIsValidCharW is also declared.